rustls-post-quantum/src/lib.rs

```diff
@@ -158,6 +158,14 @@ impl ActiveKeyExchange for Active {
         Ok(SharedSecret::from(&combined.0[..]))
     }
 
+    fn hybrid_pair(&self) -> Option<&dyn ActiveKeyExchange> {
+        Some(self.x25519.as_ref())
+    }
+
+    fn into_hybrid_pair(self: Box<Self>) -> Option<Box<dyn ActiveKeyExchange>> {
+        Some(self.x25519)
+    }
```

Comment:
It feels a little bit like these two things should be coupled together more tightly, like `hybrid_pair()` should be returning something that can act as a `&dyn ActiveKeyExchange`, but also be translated to a `Box<dyn ActiveKeyExchange>` once the decision point to use it has been reached vs asking the `ActiveKeyExchange` in two operations.

I'm not sure this is a fully baked thought yet but maybe it's something you already considered?

Reply:
The tricky part I think is avoiding consuming the original `ActiveKeyExchange` until the right time, and not breaking the API property that we want, which is that any given `ActiveKeyExchange` is only `complete()`ed at most once.

Reply:
But those sound invariants the Rust type system is pretty good at enforcing? I'm not sure I currently understand it enough to write it, but suggest you try and we can discuss the design?

Reply:
As an alternative approach, how do you both feel about this?

https://github.com/rustls/rustls/compare/main...jbp-hybrid-kx-optimisation-2

There is one decision point for whether we have a hybrid kx, but at the expense of some more surface (because we can't just call `pub_key()` and such on a different object.)

Reply:
This definitely looks like an improvement. I wonder if it would make sense to, instead of `into_hybrid()`, have `ActiveKeyExchange::complete()` yield an enum like `enum KeyExchangeResult { Complete(SharedSecret), Hybrid(Box<dyn ActiveHybridKeyExchange>) }` so that control over whether hybrid agreement is used rests with the trait implementer rather than protocol code?

Also `component` feels a little generic for the `ActiveHybridKeyExchange` methods, do we want "hybrid" to be fully generic or maybe clarify the purpose of "classical" vs "post_quantum"? And am I correct in thinking that the order doesn't matter?

@durumcrustulum not sure if you have time for this kind of thing, but perhaps you have Opinions?

Reply:
> This definitely looks like an improvement. I wonder if it would make sense to, instead of `into_hybrid()`, have `ActiveKeyExchange::complete()` yield an enum like `enum KeyExchangeResult { Complete(SharedSecret), Hybrid(Box<dyn ActiveHybridKeyExchange>) }` so that control over whether hybrid agreement is used rests with the trait implementer rather than protocol code?

I don't think that works, because the decision is actually taken by the remote server; and that controls whether we call `complete()`, or `component_complete()` (and these are the operations which need a 100% watertight at-most-once guarantee). Maybe that will be clearer once I integrate this into the protocol parts of the crate.

(That's ignoring the desire I have to avoid breaking changes to `ActiveKeyExchange` for what remains a performance enhancement.)

Reply:
> I don't think that works, because the decision is actually taken by the remote server; and that controls whether we call `complete()`, or `component_complete()` (and these are the operations which need a 100% watertight at-most-once guarantee). Maybe that will be clearer once I integrate this into the protocol parts of the crate.

Ahh, yeah, that makes sense then.

Reply:
Ah, damn, this won't work after all.

The main issue is `into_hybrid()` consumes the receiver object, which makes it not useful in the not-hybrid case.

Changing the return type to return the original receiver (ie, `Result<Box<dyn ActiveHybridKeyExchange>, Box<dyn ActiveKeyExchange>>`) doesn't work because `self: Box<Self>` cannot become a `Box<dyn ActiveKeyExchange>` without `Self: Sized`, and `Self: Sized` means we can't call it on a trait object.

Back to the drawing board...